crates/common/src/runtime_client.rs

```diff
@@ -0,0 +1,211 @@
+        match self {
+            RuntimeClientError::ProviderError(e) => e.as_serde_error(),
+            _ => None,
+        }
+    }
+}
+
+impl std::fmt::Display for RuntimeClientError {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        write!(f, "{self:?}")
+    }
+}
+
+impl From<RuntimeClientError> for ProviderError {
+    fn from(src: RuntimeClientError) -> Self {
+        match src {
+            RuntimeClientError::ProviderError(err) => err,
+            RuntimeClientError::LockError => {
+                ProviderError::CustomError("Failed to lock the client".to_string())
+            }
+        }
+    }
+}
+
+/// A provider that connects on first request allowing handling of different provider types at
+/// runtime
+#[derive(Debug, Error)]
+pub struct RuntimeClient {
+    client: Arc<RwLock<Option<InnerClient>>>,
+    url: Url,
+    max_retry: u32,
+    timeout_retry: u32,
+    initial_backoff: u64,
+    timeout: Duration,
+    /// available CUPS
+    compute_units_per_second: u64,
+}
+
+impl ::core::fmt::Display for RuntimeClient {
+    fn fmt(&self, f: &mut ::core::fmt::Formatter<'_>) -> ::core::fmt::Result {
+        write!(f, "RuntimeClient")
+    }
+}
+
+impl RuntimeClient {
+    /// Creates a new dynamic provider from a URL
+    pub fn new(
+        url: Url,
+        max_retry: u32,
+        timeout_retry: u32,
+        initial_backoff: u64,
+        timeout: Duration,
+        compute_units_per_second: u64,
+    ) -> Self {
+        Self {
+            client: Arc::new(RwLock::new(None)),
+            url,
+            max_retry,
+            timeout_retry,
+            initial_backoff,
+            timeout,
+            compute_units_per_second,
+        }
+    }
+
+    async fn connect(&self) -> Result<InnerClient, RuntimeClientError> {
+        match self.url.scheme() {
+            "http" | "https" => {
+                let client = reqwest::Client::builder()
+                    .timeout(self.timeout)
+                    .build()
+                    .map_err(|e| RuntimeClientError::ProviderError(e.into()))?;
+                let provider = Http::new_with_client(self.url.clone(), client);
+
+                #[allow(clippy::box_default)]
+                let provider = RetryClientBuilder::default()
+                    .initial_backoff(Duration::from_millis(self.initial_backoff))
+                    .rate_limit_retries(self.max_retry)
+                    .timeout_retries(self.timeout_retry)
+                    .compute_units_per_second(self.compute_units_per_second)
+                    .build(provider, Box::new(HttpRateLimitRetryPolicy));
+                Ok(InnerClient::Http(provider))
+            }
+            "ws" | "wss" => {
+                let client = Ws::connect(&self.url.to_string())
+                    .await
+                    .map_err(|e| RuntimeClientError::ProviderError(e.into()))?;
+
+                Ok(InnerClient::Ws(client))
+            }
+            "file" => {
```

Comment:
are we sure this is a infallible way to detect ipc? i'm unsure—couldn't it be just a path?

Reply:
Can definitely change it to just be a path, but will need some modification in ProviderBuilder as plain `path` is not supported by `intoUrl.

Reply:
@Evalir rebased on master, which added HTTP auth to the provider. So needed to make a few other changes. 

Also made anvil reuse the provider for tests to increase test coverage. Added a few tests that explicitly test the websocket provider against Alchemy as well.